File: platforms/mac/ShotPaste/Automation/MCP/ShotPasteMCPProtocol.swift
//
//  ShotPasteMCPProtocol.swift
//  ShotPaste
//
//  Minimal MCP 2025-11-25 JSON-RPC implementation for ShotPaste tools.
//

import Foundation

enum ShotPasteMCPProtocolOutput {
  case response([String: Any])
  case acceptedNotification
}

@MainActor
struct ShotPasteMCPProtocol {
  static let latestProtocolVersion = "2025-11-25"
  static let supportedProtocolVersions = Set([
    latestProtocolVersion,
    "2025-06-18",
    "2025-03-26",
  ])

  private let execute: (ShotPasteAutomationCommand) -> ShotPasteAutomationResult
  private let status: () -> ShotPasteAutomationResult
  private let serverVersion: () -> String

  init(
    execute: @escaping (ShotPasteAutomationCommand) -> ShotPasteAutomationResult,
    status: @escaping () -> ShotPasteAutomationResult,
    serverVersion: @escaping () -> String = {
      Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
    }
  ) {
    self.execute = execute
    self.status = status
    self.serverVersion = serverVersion
  }

  func handleMessage(_ data: Data) -> ShotPasteMCPProtocolOutput {
    let message: Any
    do {
      message = try JSONSerialization.jsonObject(with: data)
    } catch {
      return .response(Self.errorResponse(id: nil, code: -32700, message: "Parse error"))
    }

    guard let request = message as? [String: Any], request["jsonrpc"] as? String == "2.0" else {
      return .response(Self.errorResponse(id: nil, code: -32600, message: "Invalid Request"))
    }

    guard let method = request["method"] as? String else {
      // The server never sends JSON-RPC requests, so a well-formed response can
      // be safely acknowledged without additional processing.
      if request["id"] != nil, request["result"] != nil || request["error"] != nil {
        return .acceptedNotification
      }
      return .response(Self.errorResponse(id: request["id"], code: -32600, message: "Invalid Request"))
    }

    let id = request["id"]
    if id == nil {
      return handleNotification(method: method)
    }

    switch method {
    case "initialize":
      return .response(initializeResponse(id: id, params: request["params"]))
    case "ping":
      return .response(Self.successResponse(id: id, result: [:]))
    case "tools/list":
      return .response(Self.successResponse(id: id, result: ["tools": Self.tools]))
    case "tools/call":
      return .response(callToolResponse(id: id, params: request["params"]))
    default:
      return .response(Self.errorResponse(id: id, code: -32601, message: "Method not found"))
    }
  }

  private func handleNotification(method: String) -> ShotPasteMCPProtocolOutput {
    switch method {
    case "notifications/initialized", "notifications/cancelled":
      .acceptedNotification
    default:
      // JSON-RPC notifications never receive a response, including unknown
      // notification methods.
      .acceptedNotification
    }
  }

  private func initializeResponse(id: Any?, params: Any?) -> [String: Any] {
    guard
      let params = params as? [String: Any],
      let requestedVersion = params["protocolVersion"] as? String,
      params["capabilities"] is [String: Any],
      params["clientInfo"] is [String: Any]
    else {
      return Self.errorResponse(id: id, code: -32602, message: "Invalid initialize parameters")
    }

    let negotiatedVersion = Self.supportedProtocolVersions.contains(requestedVersion)
      ? requestedVersion
      : Self.latestProtocolVersion
    return Self.successResponse(
      id: id,
      result: [
        "protocolVersion": negotiatedVersion,
        "capabilities": [
          "tools": ["listChanged": false],
        ],
        "serverInfo": [
          "name": "shotpaste-macos",
          "title": "ShotPaste for macOS",
          "version": serverVersion(),
          "description": "Local automation tools for the running ShotPaste application.",
        ],
        "instructions": "Use these tools to operate the local ShotPaste UI. Capture commands remain visible and user-controlled.",
      ]
    )
  }

  private func callToolResponse(id: Any?, params: Any?) -> [String: Any] {
    guard
      let params = params as? [String: Any],
      let name = params["name"] as? String
    else {
      return Self.errorResponse(id: id, code: -32602, message: "Invalid tools/call parameters")
    }

    let arguments: [String: Any]
    if let suppliedArguments = params["arguments"] {
      guard let suppliedArguments = suppliedArguments as? [String: Any] else {
        return Self.errorResponse(id: id, code: -32602, message: "Tool arguments must be an object")
      }
      arguments = suppliedArguments
    } else {
      arguments = [:]
    }

    let result: ShotPasteAutomationResult
    switch name {
    case "shotpaste.get_status":
      guard arguments.isEmpty else {
        return Self.toolResultResponse(
          id: id,
          result: .failure("shotpaste.get_status does not accept arguments.")
        )
      }
      result = status()
    case "shotpaste.start_capture":
      guard
        Set(arguments.keys).isSubset(of: ["mode"]),
        let rawMode = arguments["mode"] as? String,
        let mode = ShotPasteAutomationCaptureMode(rawValue: rawMode.lowercased())
      else {
        return Self.toolResultResponse(
          id: id,
          result: .failure("mode must be screenshot, scrolling, or recording.")
        )
      }
      result = execute(.startCapture(mode))
    case "shotpaste.cancel_capture":
      guard arguments.isEmpty else {
        return Self.toolResultResponse(
          id: id,
          result: .failure("shotpaste.cancel_capture does not accept arguments.")
        )
      }
      result = execute(.cancelCapture)
    case "shotpaste.open_history":
      guard Set(arguments.keys).isSubset(of: ["filter"]) else {
        return Self.toolResultResponse(
          id: id,
          result: .failure("shotpaste.open_history only accepts filter.")
        )
      }
      let filter: ShotPasteAutomationHistoryFilter?
      if let rawFilter = arguments["filter"] as? String {
        guard let parsedFilter = ShotPasteAutomationHistoryFilter(rawValue: rawFilter.lowercased()) else {
          return Self.toolResultResponse(
            id: id,
            result: .failure("filter must be all, screenshot, scrolling, recording, or clipboard.")
          )
        }
        filter = parsedFilter
      } else {
        filter = nil
      }
      result = execute(.openHistory(filter))
    case "shotpaste.open_settings":
      guard Set(arguments.keys).isSubset(of: ["tab"]) else {
        return Self.toolResultResponse(
          id: id,
          result: .failure("shotpaste.open_settings only accepts tab.")
        )
      }
      let tab: PreferencesTab?
      if let rawTab = arguments["tab"] as? String {
        guard let parsedTab = ShotPasteAutomationCommand.preferencesTab(named: rawTab) else {
          return Self.toolResultResponse(
            id: id,
            result: .failure(
              "tab must be general, capture, quick-access, history, agent, shortcuts, permissions, or advanced."
            )
          )
        }
        tab = parsedTab
      } else {
        tab = nil
      }
      result = execute(.openSettings(tab))
    case "shotpaste.control_recording":
      guard
        Set(arguments.keys).isSubset(of: ["action"]),
        let rawAction = arguments["action"] as? String,
        let action = ShotPasteAutomationRecordingAction(rawValue: rawAction.lowercased())
      else {
        return Self.toolResultResponse(
          id: id,
          result: .failure("action must be pause, resume, or stop.")
        )
      }
      result = execute(.controlRecording(action))
    default:
      return Self.errorResponse(id: id, code: -32602, message: "Unknown tool: \(name)")
    }

    return Self.toolResultResponse(id: id, result: result)
  }

  private static func successResponse(id: Any?, result: [String: Any]) -> [String: Any] {
    [
      "jsonrpc": "2.0",
      "id": id ?? NSNull(),
      "result": result,
    ]
  }

  private static func errorResponse(id: Any?, code: Int, message: String) -> [String: Any] {
    [
      "jsonrpc": "2.0",
      "id": id ?? NSNull(),
      "error": [
        "code": code,
        "message": message,
      ],
    ]
  }

  private static func toolResultResponse(
    id: Any?,
    result: ShotPasteAutomationResult
  ) -> [String: Any] {
    let structuredContent: [String: Any] = [
      "ok": result.isSuccess,
      "message": result.message,
      "state": result.state,
    ]
    let text = serializedJSONString(structuredContent)
    return successResponse(
      id: id,
      result: [
        "content": [
          ["type": "text", "text": text],
        ],
        "structuredContent": structuredContent,
        "isError": !result.isSuccess,
      ]
    )
  }

  private static func serializedJSONString(_ object: [String: Any]) -> String {
    guard
      JSONSerialization.isValidJSONObject(object),
      let data = try? JSONSerialization.data(withJSONObject: object, options: [.sortedKeys]),
      let string = String(data: data, encoding: .utf8)
    else {
      return "{\"ok\":false,\"message\":\"Unable to serialize tool result.\",\"state\":{}}"
    }
    return string
  }

  private static let resultOutputSchema: [String: Any] = [
    "type": "object",
    "properties": [
      "ok": ["type": "boolean"],
      "message": ["type": "string"],
      "state": [
        "type": "object",
        "additionalProperties": ["type": "string"],
      ],
    ],
    "required": ["ok", "message", "state"],
    "additionalProperties": false,
  ]

  private static let tools: [[String: Any]] = [
    tool(
      name: "shotpaste.get_status",
      title: "Get ShotPaste status",
      description: "Read the current One Shot, scrolling capture, recording, and history UI state.",
      inputSchema: emptyInputSchema,
      readOnly: true,
      idempotent: true
    ),
    tool(
      name: "shotpaste.start_capture",
      title: "Start ShotPaste capture",
      description: "Open the visible One Shot selection UI with screenshot, scrolling, or recording preselected.",
      inputSchema: [
        "type": "object",
        "properties": [
          "mode": [
            "type": "string",
            "enum": ShotPasteAutomationCaptureMode.allCases.map(\.rawValue),
            "description": "The One Shot mode to preselect.",
          ],
        ],
        "required": ["mode"],
        "additionalProperties": false,
      ],
      readOnly: false,
      idempotent: false
    ),
    tool(
      name: "shotpaste.cancel_capture",
      title: "Cancel ShotPaste capture",
      description: "Cancel the active One Shot selection session, if one exists.",
      inputSchema: emptyInputSchema,
      readOnly: false,
      idempotent: true
    ),
    tool(
      name: "shotpaste.open_history",
      title: "Open ShotPaste history",
      description: "Open the full ShotPaste history UI, optionally with a product filter.",
      inputSchema: [
        "type": "object",
        "properties": [
          "filter": [
            "type": "string",
            "enum": ShotPasteAutomationHistoryFilter.allCases.map(\.rawValue),
            "description": "Omit this value to use the user's configured default filter.",
          ],
        ],
        "additionalProperties": false,
      ],
      readOnly: false,
      idempotent: true
    ),
    tool(
      name: "shotpaste.open_settings",
      title: "Open ShotPaste settings",
      description: "Open ShotPaste settings at an optional tab.",
      inputSchema: [
        "type": "object",
        "properties": [
          "tab": [
            "type": "string",
            "enum": [
              "general", "capture", "quick-access", "history", "agent", "shortcuts", "permissions", "advanced",
            ],
          ],
        ],
        "additionalProperties": false,
      ],
      readOnly: false,
      idempotent: true
    ),
    tool(
      name: "shotpaste.control_recording",
      title: "Control ShotPaste recording",
      description: "Pause, resume, or stop the active ShotPaste recording.",
      inputSchema: [
        "type": "object",
        "properties": [
          "action": [
            "type": "string",
            "enum": ShotPasteAutomationRecordingAction.allCases.map(\.rawValue),
          ],
        ],
        "required": ["action"],
        "additionalProperties": false,
      ],
      readOnly: false,
      idempotent: true
    ),
  ]

  private static let emptyInputSchema: [String: Any] = [
    "type": "object",
    "additionalProperties": false,
  ]

  private static func tool(
    name: String,
    title: String,
    description: String,
    inputSchema: [String: Any],
    readOnly: Bool,
    idempotent: Bool
  ) -> [String: Any] {
    [
      "name": name,
      "title": title,
      "description": description,
      "inputSchema": inputSchema,
      "outputSchema": resultOutputSchema,
      "annotations": [
        "readOnlyHint": readOnly,
        "destructiveHint": false,
        "idempotentHint": idempotent,
        "openWorldHint": false,
      ],
      "execution": ["taskSupport": "forbidden"],
    ]
  }
}

File: platforms/windows/src/ShotPaste.Windows/Services/ShotPasteMcpProtocol.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace ShotPaste.Windows.Services;

internal sealed record McpAutomationResult(
    bool Ok,
    string Message,
    IReadOnlyDictionary<string, string> State)
{
    public static McpAutomationResult Failure(string message, IReadOnlyDictionary<string, string>? state = null) =>
        new(false, message, state ?? new Dictionary<string, string>());
}

internal sealed class ShotPasteMcpProtocol(
    Func<string, JsonObject, CancellationToken, Task<McpAutomationResult>> execute,
    Func<McpAutomationResult> status,
    string serverVersion)
{
    internal const string LatestProtocolVersion = "2025-11-25";
    private static readonly HashSet<string> SupportedProtocolVersions =
        [LatestProtocolVersion, "2025-06-18", "2025-03-26"];

    public async Task<string?> HandleAsync(string json, CancellationToken cancellationToken = default)
    {
        JsonObject request;
        try
        {
            request = JsonNode.Parse(json) as JsonObject
                ?? throw new JsonException("Request must be an object.");
        }
        catch (JsonException)
        {
            return Serialize(Error(null, -32700, "Parse error"));
        }

        var id = request["id"]?.DeepClone();
        if (request["jsonrpc"]?.GetValue<string>() != "2.0")
            return Serialize(Error(id, -32600, "Invalid Request"));
        if (request["method"] is not JsonValue methodValue || !methodValue.TryGetValue<string>(out var method))
            return Serialize(Error(id, -32600, "Invalid Request"));
        if (request["id"] is null) return null;

        return method switch
        {
            "initialize" => Serialize(Initialize(id, request["params"] as JsonObject)),
            "ping" => Serialize(Success(id, new JsonObject())),
            "tools/list" => Serialize(Success(id, new JsonObject { ["tools"] = ToolDefinitions().DeepClone() })),
            "tools/call" => Serialize(await CallToolAsync(id, request["params"] as JsonObject, cancellationToken)),
            _ => Serialize(Error(id, -32601, "Method not found"))
        };
    }

    private JsonObject Initialize(JsonNode? id, JsonObject? parameters)
    {
        if (parameters?["protocolVersion"] is not JsonValue versionNode ||
            !versionNode.TryGetValue<string>(out var requestedVersion) ||
            parameters["capabilities"] is not JsonObject ||
            parameters["clientInfo"] is not JsonObject)
            return Error(id, -32602, "Invalid initialize parameters");

        var negotiated = SupportedProtocolVersions.Contains(requestedVersion)
            ? requestedVersion
            : LatestProtocolVersion;
        return Success(id, new JsonObject
        {
            ["protocolVersion"] = negotiated,
            ["capabilities"] = new JsonObject { ["tools"] = new JsonObject { ["listChanged"] = false } },
            ["serverInfo"] = new JsonObject
            {
                ["name"] = "shotpaste-windows",
                ["title"] = "ShotPaste for Windows",
                ["version"] = serverVersion,
                ["description"] = "Local automation tools for the running ShotPaste application."
            },
            ["instructions"] = "Use these tools to operate the visible local ShotPaste UI. Capture remains user-controlled."
        });
    }

    private async Task<JsonObject> CallToolAsync(JsonNode? id, JsonObject? parameters, CancellationToken cancellationToken)
    {
        if (parameters?["name"] is not JsonValue nameNode || !nameNode.TryGetValue<string>(out var name))
            return Error(id, -32602, "Invalid tools/call parameters");
        if (parameters["arguments"] is not null && parameters["arguments"] is not JsonObject)
            return Error(id, -32602, "Tool arguments must be an object");
        var arguments = parameters["arguments"] as JsonObject ?? new JsonObject();

        McpAutomationResult result;
        switch (name)
        {
            case "shotpaste.get_status" when arguments.Count == 0:
                result = status();
                break;
            case "shotpaste.get_status":
            case "shotpaste.cancel_capture" when arguments.Count > 0:
                result = McpAutomationResult.Failure($"{name} does not accept arguments.", status().State);
                break;
            case "shotpaste.start_capture" when !HasOnly(arguments, "mode") ||
                                                     ReadString(arguments, "mode") is not ("screenshot" or "scrolling" or "recording"):
                result = McpAutomationResult.Failure("mode must be screenshot, scrolling, or recording.", status().State);
                break;
            case "shotpaste.open_history" when !HasOnly(arguments, "filter") ||
                                                   ReadString(arguments, "filter") is { } filter &&
                                                   filter is not ("all" or "screenshot" or "scrolling" or "recording" or "clipboard"):
                result = McpAutomationResult.Failure("filter must be all, screenshot, scrolling, recording, or clipboard.", status().State);
                break;
            case "shotpaste.open_settings" when !HasOnly(arguments, "tab"):
                result = McpAutomationResult.Failure("shotpaste.open_settings only accepts tab.", status().State);
                break;
            case "shotpaste.control_recording" when !HasOnly(arguments, "action") ||
                                                        ReadString(arguments, "action") is not ("pause" or "resume" or "stop"):
                result = McpAutomationResult.Failure("action must be pause, resume, or stop.", status().State);
                break;
            case "shotpaste.cancel_capture":
            case "shotpaste.start_capture":
            case "shotpaste.open_history":
            case "shotpaste.open_settings":
            case "shotpaste.control_recording":
                result = await execute(name, arguments, cancellationToken);
                break;
            default:
                return Error(id, -32602, $"Unknown tool: {name}");
        }
        return ToolResult(id, result);
    }

    private static bool HasOnly(JsonObject arguments, params string[] names) =>
        arguments.All(pair => names.Contains(pair.Key, StringComparer.Ordinal));

    internal static string? ReadString(JsonObject arguments, string name)
    {
        if (arguments[name] is not JsonValue value || !value.TryGetValue<string>(out var result)) return null;
        return result.Trim().ToLowerInvariant();
    }

    private static JsonObject ToolResult(JsonNode? id, McpAutomationResult result)
    {
        var state = new JsonObject(result.State.Select(pair =>
            new KeyValuePair<string, JsonNode?>(pair.Key, JsonValue.Create(pair.Value))));
        var structured = new JsonObject
        {
            ["ok"] = result.Ok,
            ["message"] = result.Message,
            ["state"] = state
        };
        return Success(id, new JsonObject
        {
            ["content"] = new JsonArray(new JsonObject
            {
                ["type"] = "text",
                ["text"] = structured.ToJsonString(new JsonSerializerOptions { WriteIndented = false })
            }),
            ["structuredContent"] = structured.DeepClone(),
            ["isError"] = !result.Ok
        });
    }

    private static JsonArray ToolDefinitions()
    {
        static JsonObject Schema(params (string Name, JsonArray Enum, bool Required)[] properties)
        {
            var props = new JsonObject();
            var required = new JsonArray();
            foreach (var property in properties)
            {
                props[property.Name] = new JsonObject { ["type"] = "string", ["enum"] = property.Enum };
                if (property.Required) required.Add(property.Name);
            }
            var schema = new JsonObject { ["type"] = "object", ["properties"] = props, ["additionalProperties"] = false };
            if (required.Count > 0) schema["required"] = required;
            return schema;
        }

        static JsonObject Tool(string name, string title, string description, JsonObject schema) => new()
        {
            ["name"] = name, ["title"] = title, ["description"] = description, ["inputSchema"] = schema
        };

        return new JsonArray
        {
            Tool("shotpaste.get_status", "Get ShotPaste status", "Read visible capture, recording, and history state.", Schema()),
            Tool("shotpaste.start_capture", "Start capture", "Open visible One Shot UI in a capture mode.",
                Schema(("mode", new JsonArray("screenshot", "scrolling", "recording"), true))),
            Tool("shotpaste.cancel_capture", "Cancel capture", "Cancel the active One Shot UI.", Schema()),
            Tool("shotpaste.open_history", "Open history", "Open history with an optional filter.",
                Schema(("filter", new JsonArray("all", "screenshot", "scrolling", "recording", "clipboard"), false))),
            Tool("shotpaste.open_settings", "Open settings", "Open an optional settings page.",
                Schema(("tab", new JsonArray("general", "capture", "quick-access", "history", "shortcuts", "appearance", "advanced"), false))),
            Tool("shotpaste.control_recording", "Control recording", "Pause, resume, or stop an active recording.",
                Schema(("action", new JsonArray("pause", "resume", "stop"), true)))
        };
    }

    private static JsonObject Success(JsonNode? id, JsonNode result) => new()
    {
        ["jsonrpc"] = "2.0", ["id"] = id?.DeepClone(), ["result"] = result
    };

    private static JsonObject Error(JsonNode? id, int code, string message) => new()
    {
        ["jsonrpc"] = "2.0",
        ["id"] = id?.DeepClone(),
        ["error"] = new JsonObject { ["code"] = code, ["message"] = message }
    };

    private static string Serialize(JsonObject value) => value.ToJsonString(new JsonSerializerOptions { WriteIndented = false });
}
